server/channels/app/authorization.go

```diff
@@ -145,50 +137,31 @@ func (a *App) SessionHasPermissionToChannels(c request.CTX, session model.Sessio
 	}
 
 	ids, err := a.Srv().Store().Channel().GetAllChannelMembersForUser(session.UserId, true, true)
-
 	var channelRoles []string
-	uniqueRoles := make(map[string]bool)
-	if err == nil {
-		for _, channelID := range channelIDs {
+	for _, channelID := range channelIDs {
+		if err == nil {
 			if roles, ok := ids[channelID]; ok {
-				for _, role := range strings.Fields(roles) {
-					uniqueRoles[role] = true
+				channelRoles = strings.Fields(roles)
+				if a.RolesGrantPermission(channelRoles, permission.Id) {
+					continue
 				}
 			}
 		}
-	}
-
-	for role := range uniqueRoles {
-		channelRoles = append(channelRoles, role)
-	}
-
-	if a.RolesGrantPermission(channelRoles, permission.Id) {
-		return true
-	}
-
-	channels, appErr := a.GetChannels(c, channelIDs)
-	if appErr != nil && appErr.StatusCode == http.StatusNotFound {
-		return false
-	}
-
-	// Get TeamIDs from channels
-	uniqueTeamIDs := make(map[string]bool)
-	for _, ch := range channels {
-		if ch.TeamId != "" {
-			uniqueTeamIDs[ch.TeamId] = true
+		channel, appErr := a.GetChannel(c, channelID)
+		if appErr != nil && appErr.StatusCode == http.StatusNotFound {
+			return false
 		}
-	}
 
-	var teamIDs []string
-	for teamID := range uniqueTeamIDs {
-		teamIDs = append(teamIDs, teamID)
-	}
-
-	if appErr == nil && len(teamIDs) > 0 {
-		return a.SessionHasPermissionToTeams(c, session, teamIDs, permission)
+		if appErr == nil && channel.TeamId != "" {
+			if !a.SessionHasPermissionToTeam(session, channel.TeamId, permission) {
```

Comment:
Is this correct? If the session doesn't have permission to the team, shouldn't we fallback and check the session user permissions in L160?

cc @jupenur

Reply:
@esarafianou

Reply:
I will add a comment in the new code area for this, but I don't think this check makes sense.